### ISSUE TITLE
Moving methods called on owned type parameters is problematic

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#618]
### Please describe the bug

If a variable is defined as an owned type parameter, we allow moving methods to be called on such variables. This is wrong, because if the _value_ assigned to the variable is a reference, then moving methods shouldn't be available. 

This is similar to mutable methods only being available if the type parameter defines the `mut` requirement. I don't want to introduce another such modifier though, and in fact would like to find a way to get rid of it entirely.

### Related issues

- https://github.com/inko-lang/inko/issues/639

### Please list the exact steps necessary to reproduce the bug

Run the following:

```inko
trait Example {
  fn move example
}

class Thing {}

impl Example for Thing {
  fn move example {}
}

fn example[I: Example](value: I) {
  value.example
}

class async Main {
  fn async main {
    example(Thing {})     # OK, because Thing is moved into the argument
    example(ref Thing {}) # not OK, references don't allow moving methods
    example(mut Thing {}) # same
  }
}
```

This panics with the following:

```
Stack trace (the most recent call comes last):
  /var/home/yorickpeterse/Downloads/test.inko:18 in main.Main.main
  /var/home/yorickpeterse/Downloads/test.inko:12 in main.example
  /var/home/yorickpeterse/Downloads/test.inko:8 in main.Thing.example
  /var/home/yorickpeterse/Downloads/test.inko:5 in main.Thing.$dropper
Process 'Main' (0x1ec1190) panicked: can't drop a value of type 'Thing' as it still has 1 reference(s)
```

### Operating system

Fedora Silverblue 38

### Inko version

0.13.1

### Rust version

1.72.1

**Comment by yorickpeterse** [comment]
The issue here is that the ownership of type parameters defaults to "infer" instead of "owned", and there's no way to specify "I need an owned value", or "I don't care about ownership" in case we were to default to "owned".

From the perspective of consistency with regular types, defaulting to "owned" makes the most sense. This however means that if you _don't_ care about ownership, we'd need new syntax for that. A hypothetical setup would be something like this:

```inko
fn foo[T](value: T) {}     # only accepts owned values
fn foo[T](value: any T) {} # owned, ref and mut are allowed
```

That is: `any T` means "I don't care about the ownership, and I'm OK with the restrictions that brings". Such values would essentially be treated as `ref T` in terms of what operations they allow.

What I'm not sure of is how annoying this might be. For collection such as `Array` and `Map`, we'll probably end up with `any T` in quite a few places. In addition, it's easy to just use `T` by accident thinking it covers all ownerships, until somebody reports an issue saying they can't use e.g. a `ref T`, and now you have to refactor a ton of code.

**Comment by yorickpeterse** [comment]
Possible options we can take:

1. Introduce `any T` to replace `T`, and `T` means "owned T".
2. Keep `T` as "infer T", and introduce `own T` to mean "owned T" (the inverse of the previous option).
3. Infer according to use: if a type is `T` and we call a moving or mutating method, record that somewhere. The method is then only available if `T` is owned (for moving methods), or `T: mut` for mutating methods.

Option three seems interesting, and would remove the need for explicit annotations, but I'm not sure how far we can take such an implementation. For example, `Array.opt_mut` doesn't actually mutate the `T` values in `Array`, so I'm not sure how we'd ensure it's not available for e.g `Array[ref Foo]`.

**Comment by yorickpeterse** [comment]
While I like the idea of relying on inference, I believe this to bring some problems with it. For example, if the body of `foo` depends on how `bar` is inferred, we'd have to process the body of `bar` first. This means we'd require multiple type-checking passes, depending on the dependencies between methods. We'd also run into problems with circular dependencies.

I'm currently inclined to take the following path:

- We introduce `T: move` similar to `T: mut`, restricting the possible types to `Something` and `uni Something`
- Moving methods can only be called on generic type parameters if they have the `move` requirement
- `T: move` implies/also adds `T: mut`
- If a method signature includes `mut T`, we automatically turn `T` into `T: mut` (by lazily adding a bound in the form `T: mut`). This way you'd just write `fn foo[T](value: mut T)` instead of `fn foo[T: mut](value: mut T)`. There won't be an equivalent for `T: move`, at least for now

**Comment by yorickpeterse** [comment]
In light of https://github.com/inko-lang/inko/issues/639, I'm leaning more towards making `T` only accept owned values, and introduce `any T` for when you don't care. This is more consistent with `Name` meaning "an owned `Name`", which should hopefully be easier to wrap your head around.

**Comment by yorickpeterse** [comment]
One somewhat wonky issue with `T: move` is this:

```inko
fn example[I: move](value: mut I) {

}
```

Here the `move` is is pointless, because due to `mut I` you'd never be able to pass a value that can actually be moved.

**Comment by yorickpeterse** [comment]
Thinking out loud here, so this may be a little incoherent:

`any T` feels like the wrong approach: it's too different from other languages, and it's too easy to forget adding `any`, and we'd need to change pretty much every Inko program written so far (including the standard library).

The issue is that if we do `impl Clone[Option[T]] for Option if T: Clone[T]`, and we have say `T = User`, we don't have a way of expressing that `User.clone` always returns a `User`, even for a `T = ref User`. 

While we could add `T: move`, that won't solve the case of an `impl Clone[Option[T]] for Option if T: Clone[T]` where we simply leave out the `move` requirement, so we'd still end up typing something as a `ref T` while the actual cloned value is a `T`, resulting in a leak.

The only safe way that I can think of is to make `T` owned and only owned, then come up with some sort of mechanism to also allow `ref T` and `mut T` values. That is, for `impl Clone[Option[T]] for Option if T: Clone[T]` the `T` only allows owned values, meaning `Option.clone` isn't available for `Option[ref Something]`, but _is_ available for `Option[Something]`. This is of course annoying, so we need to relax that rule somehow. For that, we could make it so that given `impl X for Y`, the implementation is also available for `uni Y`, `ref Y` and `mut Y`, but within the context of that implementation `T` is always owned. This means that `Option.clone` _can_ be used on `Option[ref Something]` and returns a `Option[Something]`.

If you instead want to return e.g `Option[ref T]` as the result of a clone, the implementation would need to be this:

```
impl Clone[Option[ref T]] for Option if T: Clone[ref T]
```

This however requires changing the type system to support multiple implementations of the same trait on the same type. Due to how methods are stored in vtables, this may result in a cascade of changes to handle name mangling and what not, such that we can have different `clone` methods on the same class. I would like to avoid this if at all possible.

**Comment by yorickpeterse** [comment]
Should we introduce `any T`, what happens here:

```inko
trait MyClone[T] {
  fn my_clone -> any T
}

class Foo {}

impl MyClone[Foo] for Foo {
  fn my_clone -> Foo {
    Foo {}
  }
}

fn example[T: MyClone[T]](value: any T) -> any T {
  value.my_clone
}

class async Main {
  fn async main {
    example(ref Foo {}).TEST
  }
}
```

Assuming `T` is compatible with `any T`, the implementation of `MyClone` for `Foo` is valid itself. But when we call `example`, we'd infer the return type as `ref Foo` instead of `Foo`, resulting in the value not being dropped and thus a leak, i.e the same behaviour as we have now.

**Comment by yorickpeterse** [comment]
Let's say that for type parameter requirements to be met, if the parameter doesn't specify ownership (i.e. just `T`), it's treated as an owned value.

Now consider this example:

```inko
trait MyClone[T] {
  fn my_clone -> T
}

class Foo {}

impl MyClone[Foo] for Foo {
  fn my_clone -> Foo {
    Foo {}
  }
}

fn example[T: MyClone[T]](value: T) -> T {
  value.my_clone
}

class async Main {
  fn async main {
    example(ref Foo {}).TEST
  }
}
```

The `value` argument is typed as `T` which has a requirement, so we treat it as an owned value. Because of that, we can't pass `ref Foo` to it.

Now let's say `example` is defined as follows:

```inko
fn example[T: MyClone[T]](value: ref T) -> T {
  value.my_clone
}
```

Now we _can_ pass a `ref Foo` to it. In this case `T` is assigned `Foo`, resulting in `T: MyClone[Foo]` or `Foo: MyClone[Foo]`, and so the return type is correctly inferred to `Foo` instead of `ref Foo`.

This however introduces a new problem, consider this definition of the `Map` type:

```inko
class pub Map[K: Hash + Equal[K], V] {
  ...
}
```

Following the above rules, the keys of a `Map` are no longer allowed to be references, which is annoying. We'd need something as complex as Rust's type system (including support for blanket implementations) to work around that, at least I think, and I really don't want such complexity in Inko's type system.

**Comment by yorickpeterse** [comment]
Part of our current problem is this:

When we check if the value passed to `T` (`ref Foo`) implements `MyClone`, we end up performing the check `MyClone[Foo] -> MyClone[ref Foo]`, because on the right we compare initially against `MyClone[T]` and `T = ref Foo`. Because we enable relaxed rules this is allowed, but it's in fact unsound in this case.

The use of relaxed rules for this is very much a hack and always has been, so maybe now is the time to come up with something better?

**Comment by yorickpeterse** [comment]
To clarify, relaxed checking allows for this:

```inko
import std.cmp.Equal

fn eq[T: Equal[T]](a: T, b: T) {}

eq(ref [10], ref [20])
```

Without relaxed rules this would fail, because `Array` implements `Equal` for `Array` and not `ref Array`.

Interesting enough, if I disable relaxed checking and run `inko check` on some of my projects (e.g. [OpenFlow](https://github.com/yorickpeterse/openflow) or [inko-markdown](https://github.com/yorickpeterse/inko-markdown)), I don't see any type errors. The same goes for the standard library's test suite.

So maybe we don't need relaxed checking as much any more as I thought? I haven't looked into the same logic being applied to type bounds though, it's possible we depend on it there more.

**Comment by yorickpeterse** [comment]
In theory we could allow `impl A for ref B`, and combine that with the removal of the relaxed checking hack. You'd still implement your traits for the owned version of the type, and those implementations are still available on values types as `B`, `mut B`, etc.

If you then pass a `B` to a `T: A`, it uses the owned implementation. If instead you passed a `ref B`, it would use the implementation of `for ref B`.

This does require some name mangling changes and virtual dispatch hashing changes to take that into account I think.

With that said, I'm not entirely sure what the use case would be of `impl A for ref B`. It _is_ a thing in Rust, but even there I think I only needed it once or twice in the last 8 years, and in both cases it was a dubious case at best.

**Comment by yorickpeterse** [comment]
I remembered one case where the current setup does benefit:

```inko
class async Main {
  fn async main {
    let map = Map.new
    let key = [10]

    map.set(ref key, 42)
  }
}
```

If we remove the relaxed checking hack, this code no longer compiles because of what I outlined in https://github.com/inko-lang/inko/issues/618#issuecomment-1815471895.

Rust handles this using blanket trait implementations, such that `Hash` and `Equal` are implemented for any `&T` if `T` implements `Hash` and `Equal`.

I however want to avoid adding that kind of complexity, while at the same time I don't want to require users to do `impl X for Y` _and_ `impl X for ref Y` for every implementation just to allow code like this.

**Comment by yorickpeterse** [comment]
Some more thoughts:

We don't need `impl X for ref Y` and the likes, rather what we need is to change how type parameters are evaluated in the trait bit. That is, for `impl X[Y] for Z`, we'd change the `Y` to always produce an owned counterpart of the type assigned to it. So if `Y = ref Foo`, we'd end up with `X[Foo]` not `X[ref Foo]`. This way implementations such as `impl Clone[Option[T]] for Option` result in the correct types.

Next, we'd get rid of the relaxed rule hack and instead ensure that we drop the ownership of the trait comparison, so instead of e.g. `ref SomeTrait -> SomeTrait` we just compare `SomeTrait -> SomeTrait`. The ownership isn't relevant for the implementation as you can't write `impl ref SomeTrait ...` anyway, and we already check for the correct ownership when passing values (e.g. as an argument).

This should result in implementations being available regardless of whether you have a `ref T`, `T` or `mut T`, without requiring extensive changes.

This approach does bring two questions/challenges:

First, it's a little confusing (potentially) that a type parameter `T` in most contexts resolves to whatever type is assigned to it including its ownership (as in `T` becomes `ref Foo` if assigned `ref Foo`), but when passed as type arguments in an implementation it always resolves into an owned value (`Foo` in this case). Based on the assumption that such inconsistencies are likely the result of oversight, I'm having doubts about this approach.

Second, I think it can lead to wrong or at least confusing behaviour elsewhere. Consider this:

```inko
trait Example[T] {
  fn example -> T
}

impl Example[T] for Option[T] {
  ...
}
```

With the above rules, given an `Option[User]` the return type of `Option.example` is `User`. The same is true for `Option[ref User]`. But depending on what the trait is used for, the user may have expected the return type to be `Option[ref User]` if the receiver is also `Option[ref User]`, because `T = ref User`.

**Comment by yorickpeterse** [comment]
To illustrate this inconsistency, the following Rust code doesn't even compile:

```rust
trait Example<T> {
    fn example(&self) -> T;
}

struct Value<T> {
    value: T,
}

impl<T> Example<T> for Value<T> {
    fn example(&self) -> T {
        self.value
    }
}

fn main() {}
```

It fails with:

```
error[E0507]: cannot move out of `self.value` which is behind a shared reference
  --> src/main.rs:11:9
   |
11 |         self.value
   |         ^^^^^^^^^^ move occurs because `self.value` has type `T`, which does not implement the `Copy` trait

For more information about this error, try `rustc --explain E0507`.
error: could not compile `playground` (bin "playground") due to previous error
```

This is in fact also true for Inko:

```inko
trait Example[T] {
  fn example -> T
}

class Value[T] {
  let @value: T
}

impl Example[T] for Value {
  fn example -> T {
    @value
  }
}

class async Main {
  fn async main {}
}
```

Which fails with:

```
/var/home/yorickpeterse/homes/arch/Downloads/test.inko:11:5 error(invalid-type): expected a value of type 'T', found 'ref T'
```

Given that we specialize generics, I _think_ it should be fine if we evaluate type parameters to owned values in implementations. The next step would be to figure out how we can generalize that so it's the same everywhere.

**Comment by yorickpeterse** [comment]
Playing around with this on the `fix-moving-methods` branch, I _think_ we can dig ourselves out of this hole by forcing type parameters to be owned within the context of trait implementations. I'm using the following sketch to play around with things:

```inko
trait pub MyClone[T] {
  fn pub my_clone -> T
}

class List[T] {
  let @values: Array[T]
}

impl MyClone[Int] for Int {
  fn pub my_clone -> Int {
    self
  }
}

impl MyClone[Array[T]] for Array if T: MyClone[T] {
  fn pub my_clone -> Array[T] {
    []
  }
}

impl MyClone[List[T]] for List if T: MyClone[T] {
  fn pub my_clone -> List[T] {
    List { @values = [] }
  }
}

class async Main {
  fn async main {
    List { @values = [ref [10]] }.my_clone.TEST
  }
}
```

Version 0.13.1 produces:

```
/var/home/yorickpeterse/homes/arch/Downloads/test.inko:29:5 error(invalid-symbol): the method 'TEST' isn't defined for type 'List[ref Array[Int]]'
```

The disgusting hack however produces:

```
/var/home/yorickpeterse/homes/arch/Downloads/test.inko:29:5 error(invalid-symbol): the method 'TEST' isn't defined for type 'List[Array[Int]]'
```

The new behaviour is correct, at least for this example, as we're returning a new array, not a reference to it.

**Comment by yorickpeterse** [comment]
The idea of making type parameters in trait implementations implicitly owned isn't sound either. Take this example:

```inko
import std.drop.Drop
import std.stdio.STDOUT

class Thing {}

impl Drop for Thing {
  fn mut drop {
    STDOUT.new.print('Thing.drop')
  }
}

trait Example[VAL] {
  fn move example -> VAL
}

class Box[BOX] {
  let @value: BOX
}

impl Example[BOX] for Box {
  fn move example -> BOX {
    @value
  }
}

class async Main {
  fn async main {
    let a = Thing {}

    Box { @value = ref a }.example
  }
}
```

Running this produces the following:

```
Thing.drop
Stack trace (the most recent call comes last):
  /var/home/yorickpeterse/homes/arch/Downloads/test.inko:30 in main.Main.main
  /var/home/yorickpeterse/homes/arch/Downloads/test.inko:4 in main.Thing.$dropper
Process 'Main' (0x55d25bac9970) panicked: can't drop a value of type 'Thing' as it still has 1 reference(s)
```

What seems to happen here is that the return type of `Box.example` is now resolved into `Thing` instead of `ref Thing`. That particular case could perhaps be fixed, but I suspect this will result in a game of whack-a-mole as a result of the admittedly odd rule of "in certain contexts, type parameters are owned".

In other words, we need something better.

**Comment by yorickpeterse** [comment]
[My comment earlier](https://github.com/inko-lang/inko/issues/618#issuecomment-1814814023) about `any T` isn't quite correct I think. The `example` method is defined as follows:

```inko
fn example[T: MyClone[T]](value: any T) -> any T {
  value.my_clone
}
```

If we pass it for example `ref Foo`, then we'd have `T: MyClone[ref Foo]`. If we then remove the current ownership hack when checking traits, the type checker would produce an error because `Foo` implements `MyClone[Foo]` and not `MyClone[ref Foo]`.

**Comment by yorickpeterse** [comment]
Pony uses something similar with its `Any T` types: https://tutorial.ponylang.io/generics/#generic-classes. Pony really was on to something.

**Comment by yorickpeterse** [comment]
Pony has something similar with its `Any` constraints: https://tutorial.ponylang.io/generics/#generic-classes

**Comment by yorickpeterse** [comment]
My current plan is as follows:

- We introduce `any T` for generic type parameters. This effectively replaces the current `T` syntax. `any T` is only valid if `T` is a type parameter, meaning `any String` is a semantical error
- `T` (as in the type parameter) becomes an _owned_ type parameter. This means `fn foo[T](val: T)` expects an _owned_ value, not e.g. a `ref T`
- We adjust the standard library and what not accordingly
- For `Option.clone` et all, we make sure owned types are used where appropriate/desired (e.g. as its return type)
- We'll leave out overloading/implementing the same trait multiple times for now. This is likely to cause friction, but it's too big of a challenge to solve now. It's also not clear yet _how_ necessary this is, as I suspect in most cases people will want to use `any T` for their generics anyway
- We remove the ownership hack when checking type parameter bounds and trait implementations

I'm not _entirely_ sure about this yet though, as I suspect we may run into some issues where e.g. `clone` won't be available for references to generic types depending on the `impl` block, so I'll need to look into that.

**Comment by yorickpeterse** [comment]
This does bring the question, what happens with this:

```inko
impl Array if T: mut {
  fn foo(value: any T) { ... }
}
```

Here the `mut` constraint would restrict `T` to being owned or `mut T`, yet `any T` would allow `T`, `mut T` and `ref T`. If `T` is owned, to accept both `T` and `mut T`, you'd need to combine `T: mut` with `any T`. This gets a little confusing to be honest.

**Comment by yorickpeterse** [comment]
Here's another challenge:

If `T` is owned, then how would we allow this:

```inko
class Box[T] {
  ...
}

trait Clone[T] {
  fn pub clone -> T
}

impl Clone[any T] for Box if T: Clone[any T] {
  fn pub clone -> T {
    ...
  }
}
```

That is, for `Clone` we don't really care if the implementation is `Clone[Foo]`, `Clone[ref Foo]`, etc, as long as the return type of `clone` is an owned value. Since `any T` can also be assigned a `ref T` or `mut T`, it can't be compatible with an (owned) `T`, so the above implementation wouldn't be legal.

To allow this, we'd have to make it so that given a trait in the form `Name[Param1, Param2, ...]`, the parameters `Param1` and so on _can_ be assigned a type of any ownership, such that e.g. `Clone[ref Foo]` and `Clone[uni Foo]` are valid. Only when _referring_ to a type parameter `T` would it be treated as an owned value.

**Comment by yorickpeterse** [comment]
Another option is to flip things around: `T` stays as-is, meaning it can be any ownership, and we introduce `move T` for the cases where we want an owned value (as in, a `T` of which ownership is _moved_). This way you'd define `Clone` like so:

```inko
trait pub Clone[T] {
  fn pub clone -> move T
}
```

This requires fewer changes and probably better captures what users want with generic type parameters most of the time. You also don't get the wonky mix of `any T` and `T: mut`, and it's similar to the idea of reintroducing one-shot/moving closures (see #347).

The semantics of `move T` would be that it takes whatever type is assigned to `T` and strips its ownership, such that `T = ref Foo` is turned into `T = Foo`.

**Comment by yorickpeterse** [comment]
Funny enough, Gel (the language that introduced the basis for Inko's ownership model) does exactly the same per [the paper](https://inko-lang.org/papers/ownership.pdf) (section "6.2 Polymorphism"), using the `T%` syntax.

**Comment by yorickpeterse** [comment]
Regarding https://github.com/inko-lang/inko/issues/618#issuecomment-1823066693:

The flaw here is that `ref Thing` is compatible with `BOX`, which should be disallowed given `ref T` isn't compatible with `T`. To make this example correct, the field should be defined as `let @value: any BOX`, at which point `fn move example` would have to be typed to return `any BOX`, solving the drop panic.

For https://github.com/inko-lang/inko/issues/618#issuecomment-1825011200:

Using `move T`, we'd be able to implement `Equal` over any ownership like so:

```inko
impl Equal[User] for User {
  fn pub ==(other: ref User) -> Bool { ... }
}

impl Equal[Array[move T]] for Array if T: Equal[move T] {
  fn pub ==(other: ref Array[move T]) -> Bool { ... }
}
```

So if we have `Array[ref User]` we'd end up with `Equal[Array[ref User]]`, which due to `move T` is turned into `Equal[Array[User]]`.

The downside is that implementing generics this way can become rather verbose as a result of having to sprinkle `move T` all over the place.

In contrast, using `any T` we'd just keep the implementation as-is:

```inko
impl Equal[Array[T]] for Array if T: Equal[T] {
  fn pub ==(other: ref Array[T]) -> Bool { ... }
}
```

Given an `Array[T = ref User]`, we'd find `Equal[Array[T]]` and resolve that to `Equal[Array[User]]`, because the `ref` is stripped off.

Both both approaches suffer from a problem: `other` is transformed from e.g. `ref Array[ref User]` to `ref Array[User]`. This changes the semantics of the value in ways that are not sound. For example, had we used `other: mut Array[T]` we'd be able to do `other.pop.unwrap`, and the returned value would always be treated as an owned value, even if in fact it's a reference.

To prevent this from happening, we'd have to make `Array[ref User]` not compatible with `Array[T]` rather than just stripping off the ownership, but in doing so we can't get rid of the ownership hack employed when checking traits and bounds.

We can combine the two and get a bit closer: `T` behaves as-is, meaning it supports any ownership. `move T` strips off ownership, and only owned values/types are compatible with it. `Equal` is then implemented for `Array` like so:

```inko
impl Equal[Array[T]] for Array if T: Equal[T] {
  fn pub ==(other: ref Array[T]) -> Bool { ... }
}
```

`Clone` is then defined and implemented for `Option` as follows:

```inko
trait pub Clone[T] {
  fn pub clone -> move T
}

impl Clone[Option[move T]] for Option if T: Clone[move T] {
  fn pub clone -> Option[move T] {
    ...
  }
}
```

This however isn't quite right either: if we have `Option[T = ref User]` we'd end up with `Clone[Option[ref User]]` which wouldn't be compatible with `Clone[Option[move User]]`. I think to make that work we'd have to make `ref T` compatible with `move T`, but only allow `move T` in return types, though I suspect that will produce issues elsewhere.

**Comment by yorickpeterse** [comment]
To make `move T` safe and useful, we'd have to make `ref T` / etc compatible with it, but _only_ allow it to be used inside return types and trait type arguments. This would allow for the following:

```inko
trait pub Clone[T] {
  fn pub clone -> move T
}

impl Clone[Option[move T]] for Option if T: Clone[move T] {
  fn pub clone -> Option[move T] {
    ...
  }
}
```

But this would be invalid:

```inko
fn foo[A](value: move A) { ... }
```

This however is a bit of an odd system, and likely still results in rather verbose/tedious trait implementations. We'd also still need a solution for the ownership hack.

**Comment by yorickpeterse** [comment]
Regarding https://github.com/inko-lang/inko/issues/618#issuecomment-1814814023:

I think my original assessment isn't entirely correct. That is, if we make it such that passing `ref Foo` to `any T` results in `T = ref Foo` instead of `T = Foo` (i.e `any` doesn't cancel out the `ref`), then we end up with `T: MyClone[ref Foo]`, but `Foo` implements `MyClone[Foo]` and not `MyClone[ref Foo]`, so the compiler would emit an error at the call site.

If we wanted the `example` method to work for `ref Foo` and the likes as well, we'd have to write this:

```inko
trait MyClone[T] {
  fn my_clone -> T
}

class Foo {}

impl MyClone[Foo] for Foo {
  fn my_clone -> Foo {
    Foo {}
  }
}

fn example[T: MyClone[T]](value: ref T) -> T {
  value.my_clone
}

class async Main {
  fn async main {
    example(ref Foo {}).TEST
  }
}
```

Here passing `ref Foo` to `ref T` results in `T = Foo` as the `ref` on both sides "cancels out", which we'd have to implement. This results in `T: MyClone[Foo]` as the constraint, and `Foo` as the return type.

**Comment by yorickpeterse** [comment]
Regarding the `Map` example from https://github.com/inko-lang/inko/issues/618#issuecomment-1814852547:

We could change the definition of `Equal` to the following:

```inko
trait pub Equal[T] {
  fn pub ==(other: T) -> Bool
}
```

Instead of this:

```inko
trait pub Equal[T] {
  fn pub ==(other: ref T) -> Bool
}
```

Then types implement `Equal` using `impl Equal[ref MyType] for MyType`, and `Map` uses `K: Hash + Equal[ref K]`. Using this approach, I think we can allow keys of any ownership, provided they are compatible with a `ref`.

**Comment by yorickpeterse** [comment]
A paper that discusses what is essentially polymorphism over qualifiers (or in our case ownership) is "[Qualifying System F<:](https://arxiv.org/pdf/2311.07480.pdf)"
